dev/tools/bin/engine_hash.dart

```diff
@@ -0,0 +1,162 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// ignore_for_file: always_specify_types
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:args/args.dart';
+import 'package:crypto/crypto.dart';
+
+enum GitRevisionStrategy {
+  mergeBase,
+  head,
+}
+
+final _hashRegex = RegExp(r'^([a-fA-F0-9]+)');
+
+final parser = ArgParser()
+  ..addOption(
+    'strategy',
+    abbr: 's',
+    allowed: ['head', 'mergeBase'],
+    defaultsTo: 'head',
+    allowedHelp: {
+      'head': 'hash from git HEAD',
+      'mergeBase': 'hash from the merge-base of HEAD and upstream/master',
+    },
+  )
+  ..addFlag('help', abbr: 'h', negatable: false);
+
+Never printHelp({String? error}) {
+  if (error != null) {
+    stdout.writeln(error);
+    stdout.writeln();
+  }
+  stdout.writeln('''
+Calculate the hash signature for the Flutter Engine
+${parser.usage}
+''');
+  exit(error != null ? 1 : 0);
+}
+
+Future<int> main(List<String> args) async {
+  final ArgResults arguments;
+  try {
+    arguments = parser.parse(args);
+  } catch (e) {
+    printHelp(error: '$e');
+  }
+
+  if (arguments.wasParsed('help')) {
+    printHelp();
+  }
+
+  final result = await engineHash(
+    (List<String> command) => Process.run(
+      command.first,
+      command.sublist(1),
+      stdoutEncoding: utf8,
+    ),
+    revisionStrategy: GitRevisionStrategy.values.byName(
+      arguments.option('strategy')!,
+    ),
+  );
+
+  if (result.error != null) {
+    stderr.writeln('Error calculating engine hash: ${result.error}');
+    return 1;
+  }
+
+  stdout.writeln(result.result);
+
+  return 0;
+}
+
+/// Returns the hash signature for the engine source code.
+Future<({String result, String? error})> engineHash(
```

Comment:
Any reason we aren't just throwing an exception and returning a `Future<String>`?